src/pudl/analysis/demand_mapping.py

```diff
@@ -1,294 +1,942 @@
-            name in dataframe and matrix)
-        axis_scale (int): axis of normalization and demand allocation
-            1 if data being multiplied to rows
-            0 if data being multiplied to columns
-        normalize (bool): normalize along the axis mentioned
+        disagg_layer (GeoDataframe): Completely disaggregated GeoDataFrame
+        by (str or list): single column or list of columns according to which
+        the constants to be allocated are mentioned (e.g. "Demand" (constant)
+        which needs to be allocated is mapped by "id". So, "id" is the `by`
+        column)
+        allocatees (str or list): single column or list of columns according to which
+        the constants to be allocated are mentioned (e.g. "Demand" (constant)
+        which needs to be allocated is mapped by "id". So, "demand" is the
+        `allocatees` column)
+        allocators (str or list): columns by which attribute is weighted and
+        allocated
+        aggregators (str or list): if empty list, the disaggregated data is
+        returned. If aggregators is mentioned, for example REEDs geometries, the
+        data is aggregated at that level.
 
     Returns:
-        pandas.DataFrame: Intersection matrix scaled by vector (row-wise or
-        column-wise)
-
+        geopandas.GeoDataFrame: Disaggregated GeoDataFrame with all the various
+        allocated demand columns, or aggregated by `aggregators`
     """
-    if axis_scale == 1:
-        unique_index = intersection_matrix.index
-        index_name = intersection_matrix.index.name
+    # Allowing for single and multiple allocators,
+    # aggregating columns and allocatees
+    if type(allocators) is not list:
+        allocators = [allocators]
+
+    if type(allocatees) is not list:
+        allocatees = [allocatees]
+
+    if type(by) is not list:
+        by = [by]
+
+    # temp_allocator is product of all allocators in the row
+    disagg_layer["temp_allocator"] = disagg_layer[allocators].product(axis=1)
+
+    # the fractional allocation for each row is decided by the multiplier:
+    # (temp_allocator/temp_allocator_agg)
+    agg_layer = (disagg_layer[by + ["temp_allocator"]]
+                 .groupby(by)
+                 .sum()
+                 .reset_index()
+                 .rename(columns={"temp_allocator": "temp_allocator_agg"}))
+
+    # adding temp_allocator_agg column to the disagg_layer
+    disagg_layer = disagg_layer.merge(agg_layer)
+    allocatees_agg = [allocatee + "_allocated" for allocatee in allocatees]
+
+    # creating new allocated columns based on the allocation factor
+    disagg_layer[allocatees_agg] = disagg_layer[allocatees].multiply(disagg_layer["temp_allocator"]
+                                                                     / disagg_layer["temp_allocator_agg"],
+                                                                     axis=0)
+
+    # grouping by the relevant columns
+    if type(aggregators) is list:
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Using type() instead of isinstance() for a typecheck.](https://app.codacy.com/manual/zaneselvans/pudl/pullRequest?prid=5664147)